rosbag2_py/src/rosbag2_py/_transport.cpp

```diff
@@ -190,27 +190,116 @@ class Player
+
+    exec_ = std::make_shared<rclcpp::executors::SingleThreadedExecutor>();
+    exec_->add_node(player_);
+    spin_thread_ = std::thread(
+      [this]() {
+        exec_->spin();
+      });
+    // Wait for the executor to start spinning to avoid race conditions
+    while (!exec_->is_spinning()) {
+      std::this_thread::sleep_for(std::chrono::milliseconds(1));
+    }
   }
 
   virtual ~Player()
   {
+    if (exec_) {
+      exec_->cancel();
+      if (spin_thread_.joinable()) {
+        spin_thread_.join();
+      }
+    }
+
     rclcpp::shutdown();
   }
 
-  static void cancel()
+  void play()
   {
-    exit_ = true;
-    wait_for_exit_cv_.notify_all();
+    if (!player_) {
+      throw std::runtime_error("Player is not initialized. Please use constructor with "
+        "storage and play options.");
+    }
+    player_->play();
   }
 
-  void play()
+  bool wait_for_playback_to_start(double timeout = -1.0)
   {
-    play_impl(false);
+    if (!player_) {
+      throw std::runtime_error("Player is not initialized. Please use constructor with "
+        "storage and play options.");
+    }
+    return player_->wait_for_playback_to_start(std::chrono::duration<double>(timeout));
   }
 
-  void burst(size_t num_messages)
+  bool wait_for_playback_to_finish(double timeout = -1.0)
   {
-    play_impl(true, num_messages);
+    if (!player_) {
+      throw std::runtime_error("Player is not initialized. Please use constructor with "
+        "storage and play options.");
+    }
+    return player_->wait_for_playback_to_finish(std::chrono::duration<double>(timeout));
+  }
+
+  void stop()
+  {
+    if (!player_) {
+      throw std::runtime_error("Player is not initialized. Please use constructor with "
+        "storage and play options.");
+    }
+    player_->stop();
+  }
+
+  void pause()
+  {
+    if (!player_) {
+      throw std::runtime_error("Player is not initialized. Please use constructor with "
+        "storage and play options.");
+    }
+    player_->pause();
+  }
+
+  void resume()
+  {
+    if (!player_) {
+      throw std::runtime_error("Player is not initialized. Please use constructor with "
+        "storage and play options.");
+    }
+    player_->resume();
+  }
+
+  bool play_next()
+  {
+    if (!player_) {
+      throw std::runtime_error("Player is not initialized. Please use constructor with "
+        "storage and play options.");
+    }
+    return player_->play_next();
+  }
+
+  size_t burst(size_t num_messages)
+  {
+    if (!player_) {
+      throw std::runtime_error("Player is not initialized. Please use constructor with "
+        "storage and play options.");
+    }
+    return player_->burst(num_messages);
+  }
+
+  void seek(int64_t time_point)
+  {
+    if (!player_) {
+      throw std::runtime_error("Player is not initialized. Please use constructor with "
+        "storage and play options.");
+    }
+    player_->seek(static_cast<rcutils_time_point_value_t>(time_point));
+  }
```

Comment:
I feel like we really have two Python `Player` classes now: the existing utility wrapper for `ros2 bag play` and the more trivial wrapper for the full-ish player API (with C++ global context init and C++ executor).

I think it would be better to split them up, but I don't know if we can change the existing API.